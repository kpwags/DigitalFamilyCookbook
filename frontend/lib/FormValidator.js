import validator from 'validator';

class FormValidator {
    static validateEmail(value) {
        if (typeof value !== 'string' || !validator.isEmail(value)) {
            return false;
        }

        return true;
    }

    static validateInvitationCode(value) {
        if (!this.validateNotEmpty(value)) {
            return {
                valid: false,
                message: 'Invitation code is required'
            };
        }

        const regEx = new RegExp('^[A-Za-z0-9_]+$');
        if (!regEx.test(value)) {
            return {
                valid: false,
                message: 'Invitation codes can only contain letters, numbers, and underscores'
            };
        }

        if (value.trim().length < 1 || value.trim().length > 20) {
            return {
                valid: false,
                message: 'Invitation codes must be between 1 and 20 characters'
            };
        }

        return {
            valid: true,
            message: ''
        };
    }

    static validateNotEmpty(value) {
        if (typeof value !== 'string' || value.trim().length === 0) {
            return false;
        }

        return true;
    }

    static validateSetLength(value, fieldName = 'value', minLength = 1, maxLength = 25) {
        if (!this.validateNotEmpty(value)) {
            return {
                valid: false,
                message: `${fieldName} is required`
            };
        }

        if (value.trim().length < minLength || value.trim().length > maxLength) {
            return {
                valid: false,
                message: `${fieldName} must be between ${minLength} and ${maxLength} characters`
            };
        }

        return {
            valid: true,
            message: ''
        };
    }

    static validateUsername(value) {
        if (!this.validateNotEmpty(value)) {
            return {
                valid: false,
                message: 'Username is required'
            };
        }

        const regEx = new RegExp('^[A-Za-z0-9_]+$');
        if (!regEx.test(value)) {
            return {
                valid: false,
                message: 'Username can only contain letters, numbers, and underscores'
            };
        }

        if (value.trim().length < 1 || value.trim().length > 20) {
            return {
                valid: false,
                message: 'Username must be between 1 and 20 characters'
            };
        }

        return {
            valid: true,
            message: ''
        };
    }

    static validatePassword(value1, value2) {
        if (typeof value1 !== 'string' || typeof value2 !== 'string' || (value1 === '' && value2 === '')) {
            return {
                valid: false,
                message: 'Password is required'
            };
        }

        if (value1.trim() !== value2.trim()) {
            return {
                valid: false,
                message: 'Passwords do not match'
            };
        }

        if (value1.trim().length < 8) {
            return {
                valid: false,
                message: 'Passwords must be at least 8 characters long'
            };
        }

        return {
            valid: true,
            message: ''
        };
    }
}

export { FormValidator };
