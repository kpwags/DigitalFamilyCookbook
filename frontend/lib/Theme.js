const Theme = {
    black: 'hsl(0, 0%, 22.4%)',
    green: 'hsl(140, 100%, 25%)',
    lightGreen: 'hsl(136, 44%, 47%)',
    paleGreen: 'hsl(136, 100%, 88%)',
    darkGreen: 'hsl(140, 100%, 15%)',
    bs: '0 12px 24px 0 hsla(0, 0%, 0%, 0.2)'
};

export { Theme };
