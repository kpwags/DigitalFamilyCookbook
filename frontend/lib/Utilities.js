import Router from 'next/router';

class Utilities {
    static setTableRowActive(tableId, rowId) {
        const table = document.getElementById(tableId);
        const tbody = table.getElementsByTagName('tbody')[0];

        // TODO: Find better way?
        // eslint-disable-next-line no-restricted-syntax
        for (const row of tbody.getElementsByTagName('tr')) {
            row.className = '';
        }

        document.getElementById(rowId).className = 'selected';
    }

    // TODO: Add Message
    static invalidateField(elementId, errorMessage = '') {
        document.getElementById(elementId).className = 'errored';
        document.getElementById(elementId).parentElement.className = 'errored';

        if (errorMessage !== '') {
            document.getElementById(`${elementId}-message`).innerText = errorMessage;
            document.getElementById(`${elementId}-message`).style.display = 'block';
        }
    }

    static invalidateFieldByClass(className) {
        const elements = document.getElementsByClassName(className);

        for (let i = 0; i < elements.length; i += 1) {
            elements[i].className = `${className}  errored`;
        }
    }

    static resetFieldByClass(className) {
        const elements = document.getElementsByClassName(className);

        for (let i = 0; i < elements.length; i += 1) {
            let elementClassName = elements[i].className;
            elementClassName = elementClassName.replace('errored', '');
            elements[i].className = elementClassName;
        }
    }

    static resetField(elementId) {
        let { className } = document.getElementById(elementId);
        className = className.replace('errored', '');
        document.getElementById(elementId).className = className;

        let parentClassName = document.getElementById(elementId).parentElement.className;
        parentClassName = parentClassName.replace('errored', '');
        document.getElementById(elementId).parentElement.className = parentClassName;
        document.getElementById(`${elementId}-message`).style.display = 'none';
        document.getElementById(`${elementId}-message`).innerText = '';
    }

    static getNextAvailableValue(arr, idx = 'key') {
        let nextValue = 0;
        arr.forEach(item => {
            if (item[idx] >= nextValue) {
                nextValue = item[idx];
            }
        });

        nextValue += 1;

        return nextValue;
    }

    static convertCookingTime(time) {
        let timeHours = 0;
        let timeMinutes = 0;

        let returnString = '';

        if (time !== null && parseInt(time, 10) > 0) {
            timeHours = Math.floor(time / 60);
            timeMinutes = time % 60;

            if (timeHours === 1) {
                returnString += `${timeHours} hour`;
            } else if (timeHours > 1) {
                returnString += `${timeHours} hours`;
            }

            if (timeHours > 0 && timeMinutes > 0) {
                returnString += ', ';
            }

            if (timeMinutes > 0) {
                returnString += `${timeMinutes} minutes`;
            }
        }

        return returnString;
    }

    static convertIDArray(arr) {
        const newArr = [];
        arr.forEach(i => {
            newArr.push(i.id);
        });

        return newArr;
    }

    static toTitleCase(str) {
        return str
            .toLowerCase()
            .split(' ')
            .map(word => {
                return word.replace(word[0], word[0].toUpperCase());
            })
            .join(' ');
    }

    static goToRecipe(id) {
        Router.push({
            pathname: '/recipe',
            query: { id }
        });
    }

    static toggleElement(elementId) {
        if (document.getElementById(elementId).style.display !== 'none') {
            document.getElementById(elementId).style.display = 'none';
        } else {
            document.getElementById(elementId).style.display = 'block';
        }
    }

    static toggleMobileMenu() {
        document.getElementById('hamburgermenu').classList.toggle('change');

        if (document.getElementById('mobilemenu').style.left === '0px') {
            document.getElementById('mobilemenu').style.left = '-250px';
        } else {
            document.getElementById('mobilemenu').style.left = '0px';
        }
    }

    static toggleHeaderMenu(menuId) {
        if (
            document.getElementById(menuId).style.display === 'none' ||
            document.getElementById(menuId).style.display === ''
        ) {
            document.getElementById(menuId).style.display = 'block';
        } else {
            document.getElementById(menuId).style.display = 'none';
        }
    }

    static hideHeaderMenu(menuId) {
        document.getElementById(menuId).style.display = 'none';
    }

    static formatDate(date) {
        const formattedDate = new Date(date);
        return formattedDate.toLocaleDateString('en-US');
    }

    static boxColors = ['gray', 'blue', 'green', 'orange', 'purple'];
}

export { Utilities };
