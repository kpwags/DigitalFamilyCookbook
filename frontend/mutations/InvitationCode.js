import gql from 'graphql-tag';

const CREATE_INVITATION_CODE_MUTATION = gql`
    mutation CREATE_INVITATION_CODE_MUTATION($code: String!) {
        createInvitationCode(code: $code) {
            id
            code
        }
    }
`;

const UPDATE_INVITATION_CODE_MUTATION = gql`
    mutation UPDATE_INVITATION_CODE_MUTATION($id: ID!, $code: String) {
        updateInvitationCode(id: $id, code: $code) {
            id
            code
        }
    }
`;

const DELETE_INVITATION_CODE_MUTATION = gql`
    mutation DELETE_INVITATION_CODE_MUTATION($id: ID!) {
        deleteInvitationCode(id: $id) {
            id
        }
    }
`;

export { CREATE_INVITATION_CODE_MUTATION, UPDATE_INVITATION_CODE_MUTATION, DELETE_INVITATION_CODE_MUTATION };
