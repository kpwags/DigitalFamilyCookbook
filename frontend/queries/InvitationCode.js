import gql from 'graphql-tag';

const ALL_INVITATION_CODES_QUERY = gql`
    query {
        invitationCodes(orderBy: code_ASC) {
            id
            code
            createdAt
        }
    }
`;

const SINGLE_INVITATION_CODE_CODE_QUERY = gql`
    query SINGLE_INVITATION_CODE_CODE_QUERY($code: String!) {
        invitationCode(where: { code: $code }) {
            id
            code
        }
    }
`;

export { ALL_INVITATION_CODES_QUERY, SINGLE_INVITATION_CODE_CODE_QUERY };
