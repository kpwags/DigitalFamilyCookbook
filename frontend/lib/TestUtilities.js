/* eslint-disable import/no-extraneous-dependencies */
import casual from 'casual';
import PropTypes from 'prop-types';
import { MockedProvider } from '@apollo/react-testing';
import { ThemeProvider } from 'styled-components';
import { Theme } from './Theme';

const MockedThemeProvider = (props) => {
    const { mocks = null, addTypename = false, children } = props;
    return (
        <ThemeProvider theme={Theme}>
            <MockedProvider mocks={mocks} addTypename={addTypename}>
                {children}
            </MockedProvider>
        </ThemeProvider>
    );
};

MockedThemeProvider.propTypes = {
    children: PropTypes.node,
    mocks: PropTypes.array,
    addTypename: PropTypes.bool,
};

const TestAdmin = () => ({
    __typename: 'User',
    id: casual.uuid,
    name: casual.name,
    username: casual.username,
    email: casual.email,
    bio: 'I am just a test admin',
    image: 'user.jpg',
    largeImage: 'user-lg.jpg',
    permissions: ['USER', 'ADMIN'],
});

const TestCategory = () => ({
    __typename: 'Category',
    id: casual.uuid,
    name: casual.word,
    createdAt: casual.date,
});

const TestDirection = () => ({
    __typename: 'Direction',
    id: casual.uuid,
    name: casual.words(),
    sortOrder: casual.integer,
});

const TestIngredient = () => ({
    __typename: 'Ingredient',
    id: casual.uuid,
    name: casual.words(),
    sortOrder: casual.integer,
});

const TestInvitationCode = () => ({
    __typename: 'InvitationCode',
    id: casual.uuid,
    code: casual.word,
    createdAt: casual.date,
});

const TestMeat = () => ({
    __typename: 'Meat',
    id: casual.uuid,
    name: casual.words(),
    createdAt: casual.date,
});

const TestUser = () => ({
    __typename: 'User',
    id: casual.uuid,
    name: casual.name,
    username: casual.word,
    email: casual.email,
    bio: 'I am just a test user',
    image: 'user.jpg',
    largeImage: 'user-lg.jpg',
    permissions: ['USER'],
});

const TestRecipe = () => ({
    __typename: 'Recipe',
    id: casual.uuid,
    name: casual.name,
    public: casual.boolean,
    source: casual.words(),
    sourceUrl: casual.url,
    time: casual.integer,
    activeTime: casual.integer,
    servings: casual.integer,
    calories: casual.integer,
    carbohydrates: casual.integer,
    fat: casual.integer,
    sugar: casual.integer,
    cholesterol: casual.integer,
    fiber: casual.integer,
    image: 'recipe.jpg',
    largeImage: 'recipe-large.jpg',
    ingredients: [TestIngredient(), TestIngredient()],
    directions: [TestDirection(), TestDirection()],
    meats: [TestMeat(), TestMeat()],
    categories: [TestCategory(), TestCategory()],
    user: TestUser(),
});

const TestRecipesList = () => [TestRecipe(), TestRecipe(), TestRecipe(), TestRecipe(), TestRecipe()];

export { MockedThemeProvider, TestAdmin, TestCategory, TestDirection, TestIngredient, TestInvitationCode, TestMeat, TestRecipe, TestRecipesList, TestUser };
