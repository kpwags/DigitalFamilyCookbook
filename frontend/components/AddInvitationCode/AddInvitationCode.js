import React, { useState } from 'react';
import { useMutation, useApolloClient } from '@apollo/react-hooks';
import debounce from 'lodash.debounce';
import PropTypes from 'prop-types';
import { Form } from '../Form/Form';
import { ErrorMessage } from '../ErrorMessage/ErrorMessage';
import { TextInput } from '../TextInput/TextInput';
import { CREATE_INVITATION_CODE_MUTATION } from '../../mutations/InvitationCode';
import { ALL_INVITATION_CODES_QUERY, SINGLE_INVITATION_CODE_CODE_QUERY } from '../../queries/InvitationCode';
import { FormValidator } from '../../lib/FormValidator';

const AddInvitationCode = props => {
    const [code, setCode] = useState('');
    const [error, setError] = useState(null);
    const [codeError, setCodeError] = useState('');
    const [saveEnabled, setSaveEnabled] = useState(true);

    const client = useApolloClient();
    const [createInvitationCode, { loading: addLoading, error: addError }] = useMutation(
        CREATE_INVITATION_CODE_MUTATION,
        {
            refetchQueries: [{ query: ALL_INVITATION_CODES_QUERY }],
            onCompleted: data => {
                setCode('');
                setCodeError('');

                if (typeof props.onDone === 'function') {
                    props.onDone({
                        result: true,
                        message: `Invitation Code '${data.createInvitationCode.code}' has been added successfully.`
                    });
                }
            }
        }
    );

    const validate = debounce(async () => {
        setSaveEnabled(false);

        const res = await client.query({
            query: SINGLE_INVITATION_CODE_CODE_QUERY,
            variables: { code }
        });

        const { valid, message } = FormValidator.validateInvitationCode(code);

        if (res.data.invitationCode !== null) {
            setCodeError('Inivation code already exists');
            setSaveEnabled(false);
        } else if (!valid) {
            setCodeError(message);
            setSaveEnabled(false);
        } else {
            setCodeError('');
            setSaveEnabled(true);
        }
    }, 350);

    const cancelAdd = e => {
        e.preventDefault();
        setCode('');
        setCodeError('');

        if (typeof props.onDone === 'function') {
            props.onDone(true, null);
        }
    };

    const validateForm = async () => {
        let isValid = true;

        const res = await client.query({
            query: SINGLE_INVITATION_CODE_CODE_QUERY,
            variables: { code }
        });

        const { valid, message } = FormValidator.validateInvitationCode(code);

        if (res.data.invitationCode !== null) {
            setCodeError('Inivation code already exists');
            isValid = false;
        } else if (!valid) {
            setCodeError(message);
            isValid = false;
        } else {
            setCodeError('');
        }

        return isValid;
    };

    return (
        <Form
            data-test="form"
            id="create-invitation-code-form"
            onSubmit={async e => {
                e.preventDefault();

                setError(null);

                const isValid = await validateForm();
                if (isValid) {
                    await createInvitationCode({ variables: { code } }).catch(err => {
                        setError(err);
                    });
                }
            }}
        >
            <ErrorMessage error={addError || error} />
            <fieldset disabled={addLoading} aria-busy={addLoading}>
                <TextInput
                    id="add-invitation-code-code"
                    name="code"
                    label="Code"
                    value={code}
                    onChange={e => {
                        setCode(e.target.value);
                    }}
                    validate={() => {
                        validate();
                    }}
                    error={codeError}
                />

                <button type="submit" aria-disabled={!saveEnabled} disabled={!saveEnabled}>
                    Sav{addLoading ? 'ing' : 'e'}
                </button>
                <button type="button" onClick={cancelAdd}>
                    Cancel
                </button>
            </fieldset>
        </Form>
    );
};

AddInvitationCode.propTypes = {
    onDone: PropTypes.func
};

export { AddInvitationCode };
