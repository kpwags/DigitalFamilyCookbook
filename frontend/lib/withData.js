import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { endpoint } from '../config';
import { LOCAL_STATE_QUERY } from '../queries/Local';

function createClient({ headers }) {
    return new ApolloClient({
        uri: endpoint,
        request: operation => {
            operation.setContext({
                fetchOptions: {
                    credentials: 'include'
                },
                headers
            });
        },
        // local data
        clientState: {
            resolvers: {
                Mutation: {
                    toggleOverlay(_, variables, { cache }) {
                        const { overlayVisible } = cache.readQuery({
                            query: LOCAL_STATE_QUERY
                        });

                        const data = {
                            data: {
                                overlayVisible: !overlayVisible
                            }
                        };

                        cache.writeData(data);

                        return data;
                    }
                }
            },
            defaults: {
                overlayVisible: false
            }
        }
    });
}

export default withApollo(createClient);
