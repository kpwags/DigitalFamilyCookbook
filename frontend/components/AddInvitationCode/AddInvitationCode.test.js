import { render, fireEvent, act, waitFor } from '@testing-library/react';
import { ApolloProvider } from '@apollo/react-hooks';
import { createMockClient } from 'mock-apollo-client';
import { CREATE_INVITATION_CODE_MUTATION } from '../../mutations/InvitationCode';
import { ALL_INVITATION_CODES_QUERY, SINGLE_INVITATION_CODE_CODE_QUERY } from '../../queries/InvitationCode';
import { TestInvitationCode, MockedThemeProvider } from '../../lib/TestUtilities';
import { AddInvitationCode } from './AddInvitationCode';

const invitationCode = {
    id: '0987654321',
    code: 'testcode123',
};

const mockClient = createMockClient();
const duplicateCheckMockClient = createMockClient();

// Mock the result of the login mutation
const addInvitationCodeMutationHandler = jest.fn().mockResolvedValue({
    data: {
        createInvitationCode: {
            id: invitationCode.id,
            code: invitationCode.code,
        },
    },
});

const allInvitationCodesQueryHandler = jest.fn().mockResolvedValue({
    data: {
        invitationCodes: [TestInvitationCode(), TestInvitationCode(), TestInvitationCode()],
    },
});

const singleInvitationCodeQueryHandler = jest.fn().mockResolvedValue({
    data: {
        invitationCode: {
            id: '1234567890',
            code: 'testcode',
        },
    },
});

const singleInvitationCodeEmptyQueryHandler = jest.fn().mockResolvedValue({
    data: {
        invitationCode: null,
    },
});

mockClient.setRequestHandler(CREATE_INVITATION_CODE_MUTATION, addInvitationCodeMutationHandler);
duplicateCheckMockClient.setRequestHandler(CREATE_INVITATION_CODE_MUTATION, addInvitationCodeMutationHandler);

mockClient.setRequestHandler(ALL_INVITATION_CODES_QUERY, allInvitationCodesQueryHandler);
duplicateCheckMockClient.setRequestHandler(ALL_INVITATION_CODES_QUERY, allInvitationCodesQueryHandler);

mockClient.setRequestHandler(SINGLE_INVITATION_CODE_CODE_QUERY, singleInvitationCodeEmptyQueryHandler);
duplicateCheckMockClient.setRequestHandler(SINGLE_INVITATION_CODE_CODE_QUERY, singleInvitationCodeQueryHandler);

describe('<AddInvitationCode />', () => {
    test('it renders the input', async () => {
        const { getByLabelText } = render(
            <MockedThemeProvider>
                <AddInvitationCode />
            </MockedThemeProvider>
        );

        await waitFor(() => getByLabelText(/Code/));
    });

    test('it creates an invitation code when the form is submited', async () => {
        await act(async () => {
            const { findByText, getByLabelText, getByTestId } = render(
                <ApolloProvider client={mockClient}>
                    <AddInvitationCode />
                </ApolloProvider>
            );

            await waitFor(() =>
                fireEvent.change(getByLabelText(/Code/), {
                    target: {
                        value: invitationCode.code,
                    },
                })
            );

            await waitFor(() => fireEvent.blur(getByLabelText(/Code/)));

            await findByText(/OK/);

            await waitFor(() => fireEvent.click(getByTestId(/savebutton/)));
        });

        expect(singleInvitationCodeEmptyQueryHandler).toBeCalledTimes(2);

        expect(addInvitationCodeMutationHandler).toBeCalledWith({
            code: invitationCode.code,
        });

        expect(allInvitationCodesQueryHandler).toBeCalledTimes(1);
    });

    test('it alerts the user a invitation code is required if left blank', async () => {
        const { findByText, getByTestId } = render(
            <ApolloProvider client={mockClient}>
                <AddInvitationCode />
            </ApolloProvider>
        );

        await waitFor(() => fireEvent.click(getByTestId(/savebutton/)));

        // Assert that the error message was displayed
        await findByText(/Invitation code is required/);
    });

    test('it alerts the user the code already exists if a duplicate is typed', async () => {
        await act(async () => {
            const { findByText, getByLabelText } = render(
                <ApolloProvider client={duplicateCheckMockClient}>
                    <AddInvitationCode />
                </ApolloProvider>
            );

            await waitFor(() =>
                fireEvent.change(getByLabelText(/Code/), {
                    target: {
                        value: invitationCode.code,
                    },
                })
            );

            await waitFor(() => fireEvent.blur(getByLabelText(/Code/)));

            await findByText(/Invitation code already exists/);
        });

        expect(singleInvitationCodeQueryHandler).toBeCalledTimes(1);
    });
});
