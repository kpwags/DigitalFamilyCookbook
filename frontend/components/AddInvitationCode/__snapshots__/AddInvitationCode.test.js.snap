// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddInvitationCode/> should match snapshot 1`] = `
<form
  className="Form-sc-1jw7vbt-0 jSygZL"
  data-test="form"
  id="create-invitation-code-form"
  onSubmit={[Function]}
>
  <ErrorMessage
    error={null}
  />
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <ApolloConsumer>
      <label
        htmlFor="code"
      >
        Code
        <input
          id="add-invitation-code-code"
          maxLength="20"
          name="code"
          onChange={[Function]}
          type="text"
        />
        <div
          className="error-text"
          id="add-invitation-code-code-message"
        />
      </label>
    </ApolloConsumer>
    <button
      type="submit"
    >
      Sav
      e
    </button>
    <button
      onClick={[Function]}
      type="button"
    >
      Cancel
    </button>
  </fieldset>
</form>
`;
